### ISSUE TITLE
Q3_K CPU perplexity is 8.9% worse than llama.cpp on the same file — not the dequantizer, not the harness, not the new dots

### ISSUE THREAD

**Issue by jamesburton** [jamesburton/dotLLM#501]
Found while doing the perplexity sanity check for #497. **Pre-existing and unrelated to that work** — this is a record of a gap, with the three obvious causes already eliminated.

## The number

wikitext-2, ctx 512, 20480 tokens / 40 chunks, CPU, `Llama-3.2-1B-pure` fixtures (`llama-quantize --allow-requantize --pure --output-tensor-type q8_0 --token-embedding-type q8_0` from `bartowski/Llama-3.2-1B-Instruct-GGUF` Q8_0):

| fixture | dotLLM | llama.cpp (`-c 512 --chunks 40 -ngl 0`) | delta |
|---|---|---|---|
| Q8_0 | 14.2358 +/- 0.4048 | 14.1534 +/- 0.3968 | **+0.58%** (inside bars) |
| Q2_K | 1248.46 +/- 51.11 | 1285.14 +/- 51.55 | -2.9% (inside bars) |
| **Q3_K** | **23.7644 +/- 0.7063** | **21.8266 +/- 0.6329** | **+8.9%, bars disjoint** |

## What it is not

1. **Not the harness, tokenizer or window alignment.** The Q8_0 control runs the same corpus, the same args and the same forward pass and lands within 0.6%. (The Q2_K row is weak evidence on its own — at PPL 1250 the model is near-garbage and a few percent means nothing — which is why the Q8_0 control was run.)
2. **Not the #497 packed dots.** The same build's dequantize-and-dot path (F32 weights, no activation quantization) gives **23.7029** — 0.26% from the packed result. If anything that path should be *better* than llama.cpp's integer-quantized-activation kernel, not 9% worse.
3. **Not the Q3_K dequantizer.** `DequantizeKQuantTests.Q3_K_DenseRandomBlocks_MatchLlamaCppReference` compares dense random super-blocks element-by-element against a literal transcription of `dequantize_row_q3_K` and requires exact float equality; it passes. A per-tensor sweep of the fixture backs that up — all **112** Q3_K tensors correlate **0.9854 to 0.9886** with the Q8_0 of the same base (median 0.9883), with no outlier tensor and no shape-dependent pattern. The worst are `blk.14/15.attn_{k,v}`, which is just where 3-bit quantization hurts most.

## So where is it

Unknown. Since the decoded weights match llama.cpp bit-for-bit and the harness is calibrated, the difference has to be in how those weights are consumed — something in the forward pass that Q8_0 does not exercise but Q3_K does. The Q3_K path is also the one whose element ordering was wrong until 2026-08-10 (see the note in `DequantizeQ3_KScalar`), so a second, subtler residual from that era is a reasonable place to start looking.

Suggested next step: `perplexity --dump-tokens` / `--tokens-file` to remove tokenization from the comparison entirely, then per-layer logit comparison against llama.cpp on a fixed token sequence.

Reproduce:

```
dotllm perplexity ~/.dotllm/quant-ladder/Llama-3.2-1B-pure/Llama-3.2-1B-pure-Q3_K.gguf \
  -f C:/Development/bitnet-tests/data/wikitext-2-raw/wiki.test.raw -c 512 -n 20480 -d cpu
llama-perplexity -m <same file> -f <same corpus> -c 512 --chunks 40 -ngl 0 --no-warmup
```

**Comment by jamesburton** [comment]
Cause found, fixed in CPU and CUDA, merged to `dev` (`6d2a8a9b`).

## Cause: a fast-exp approximation in the attention softmax

`FastMath.FastExp` / `ExpSumAndStore` implemented `exp` with the Schraudolph 1999 bit trick (~1-2%
relative error), consumed by the CPU attention softmax (`Attention.cs` x6, `Softmax.cs` x1). Its own
comment justified it as "errors in exp get normalized away when dividing by the sum". That is true of
the **mean** only — the error is relative and per-element, so it survives normalization as a
reweighting of the mixture each head computes. Free on a full-precision model, which is all that was
ever checked; it costs real quality on a heavily quantized one.

Paired per-chunk, accurate minus approximate, scored on llama.cpp's exact token stream:

| fixture | dNLL | t |
|---|---|---|
| **Q3_K** | **-0.01724 +/- 0.00185 nats (-1.71% PPL)** | **-9.3** |
| Q8_0 (control) | +0.00042 +/- 0.00059 | +0.7 |

Gap to llama.cpp on Q3_K: **+2.03% (t=+5.7) -> +0.28% (t=+0.8, not significant)**. Q8_0: -0.04% -> +0.00%.

## This issue's premise "not the harness" was wrong

The corpus is CRLF and `llama-perplexity` on Windows strips the CRs through MSVC text mode — **501 of
the 512 tokens in chunk 0 differed**, so the original "+8.9%, bars disjoint" and its "+0.58% Q8_0
control" were computed on non-identical streams whose paired per-chunk sd (~0.2 nats) is an order of
magnitude larger than the effect. Filed and fixed separately as **#506**. The numbers above are from
the corrected comparison.

## Cross-backend, per CLAUDE.md

The same approximation was duplicated in CUDA, with a comment requiring the constants stay in sync
with `FastMath.cs`:

| file | sites |
|---|---|
| `native/kernels/attention_f32.cu` | 20 calls + helper + 3 defines |
| `native/kernels/attention_flash_mma_decode_gqa_split.cu` | **5 calls + helper + 3 defines** — not in the original scoping, found by sweeping |

Verified rather than assumed clean: `attention.cu` already used `expf`; **Vulkan — 12 attention
shaders, all GLSL `exp()`**; **HIP — only `rmsnorm.hip` exists, no attention path**.
`attention_flash_mma.cu`'s `__expf` was deliberately left: a hardware intrinsic at ~1e-5 relative, not
this defect.

## Validation on the T5500 (RTX 3060)

PTX regenerated with CUDA 12.8 / VS2022, arch preserved (`attention_f32` sm_75,
`attention_flash_mma_decode_gqa_split` sm_86 per `ARCH_86`), both `.version 8.7`. The Schraudolph
constant `0f4E7DE250` went **41 -> 0** and **30 -> 0**.

```
attention parity (unit)         61 tests: 59 passed, 0 failed, 2 skipped
attention parity (integration)   5 tests:  2 passed, 0 failed, 3 skipped
```

Before the PTX was regenerated the same filter failed 3 tests — CPU accurate against CUDA still
approximate. That drift appearing and then disappearing is the cleanest confirmation that the two
halves were the same defect.

## Left for a follow-up, deliberately

Tolerances calibrated around the shared approximation are now loose and were **not** touched, because
tightening is a measurement, not a guess. Strongest candidates:
`CudaAttentionSinksKernelTests` test 2 (`abs=rel=1.5e-2`, calibrated entirely on the
CPU-exact-vs-CUDA-approximate mismatch that no longer exists), `AttentionF32ParityTests` (`5e-3`),
CPU-side `TiledAttentionTests` (widened 1e-4 -> 2e-2 explicitly "to account for fast approximate exp").

Note `DOTLLM_FAST_EXP=1` is now **CPU-only** — CUDA ships precompiled PTX with no equivalent lever — so
setting it makes the backends diverge by design. Do not run cross-backend parity with it set.
